[Experimental] Create a siem agent and entity analytics tool

machadoum opened the pull request:
## Summary

This is an experimental feature designed to answer as many security solution questions as possible (from the threat hunting prompts spreadsheet) using the entity analytics knowledge tool approach.

The approach involves building a tool that "teaches" the agent the entity analytics domain and its data, enabling it to answer the questions using platform tools.

Questions it can answer at the moment:
* Which users have the highest risk scores?
* Show me how John's risk score has changed over the last 90 days
* Which 10 users have the highest risk scores right now?
* Show me entities with anomalous behavior in the last 24h
* Show users who downloaded unusually large data
* Which service accounts have unusual access patterns?
* Show users logged in from multiple locations
* Are there connections suggesting lateral movement?
* Show accounts performing unusual administrative actions
* Which users uploaded data to external domains?
* Show unusual access attempts to privileged accounts
* Show me users with suspicious login patterns
* Show me entities with anomalous behavior in the last 24 hours
* Which accounts have downloaded more than 1GB this week?
* Is anyone accessing sensitive data from new locations?
* Are there any unusual access patterns after hours?

*** Risk score questions require the risk engine to be enabled and running
*** Anomalies question requires the right ML job to be running and producing relevant data


<hr >

## Architecture

![Threat Hunting Natural Language Agent](https://github.com/user-attachments/assets/c6d94a9c-1ebf-4993-8450-80cba1e1e9c8)

**This PR contains:** 
* One new tool (Entity Analytics)
* Five subTools (Anomalies, Asset criticality, Entity Store, Risk Score and Privileged User Monitoring)
* New feature flag `securitySolution.naturalLanguageThreatHuntingEnabled`

### The Entity Analytics Tool
It is responsible for generating Entity Analytics queries and providing domain-specific knowledge.  
These are the Tool parameters:
```ts
z.object({

// defines which entity the agent needs information about. Each entity type could be stored on a different index or require a different filter. However, this helps improve query accuracy, but it also prevents us from generating queries for multiple entity types.
  entityType: EntityTypeZod.describe('The type of entity to search for.'),

//  Defines which subtool should be called
  domain: z    
    .enum([
      'risk_score',
      'asset_criticality',
      'entity_store',
      'privileged_user_monitoring',
      'anomaly_detection',
    ])
    .describe('The domain of entity analytics to retrieve information about.')
    .optional(),

  prompt: z.string().describe('The prompt or question that calling this tool will help to answer.'),

// This helps preserve filters on long conversations with follow-up queries.
  queryExtraContext: z
    .string()
    .describe('Information from previous chat messages like an ESQL filter that should be used.'),

  informationOnly: z
    .boolean()
    .describe('If true, only provide information without generating ESQL queries.'),
})
```

The Entity analytics tool calls the right subdomain, which returns domain-specific data as text. Then it calls the generate-ESQL-query tool directly and provides the prompt and the domain metadata as context for the query generation.
Then the entity analytics tool returns the generated query, which the agent will execute.


### The subTools
The name subTools is temporary. It represents a subdomain of entity analytics. The subTool is simply a function that receives dependencies (Kibana client, Elasticsearch client, etc.) and returns text. The text contains metadata for the indices that the subdomain uses that will be used by the generate ESQL tool. Example of the type of information it returns:
```
Query index TEST* for information about the domain TEST, filter by test.field == {username} when returning data about a single user...
```

***We could have promoted each subTool to an agent builder tool, but that could dramatically increase the number of tools an agent has access to, and there is already a limit of 21 tools per agent.***

The dependencies enable the subTool to search ES data and reuse Entity Analytics code via Data clients, Factories, and helpers, as well as other Kibana plugins. The `anomalies` subTool uses the ML plugins to query all enabled ML jobs; it also makes a follow-up LLM call to filter these jobs according to the prompt. We could even do something as complex as creating a LangChain flow that makes several LLM calls. 

<hr />

### Known limitation
* The evaluation test does not run on CI
* No graceful privileges verification. (We delegate the privileges check to plugins and tools, which return error messages and then the LLM interprets them)
* This PR was tested using GPT 4.1 model
* No anonymisation
* We do not control how the agent builder displays the answer
* When the questions can't be answered by the entity analytics tool, the agent builder tries to answer the question using the security solution dataview. However, it often resulted in an incomplete or inaccurate answer.
* [**Dependency**] When answering questions that use multiple indices (security solution dataview), the agent builder will throw an error due to this check [here](https://github.com/machadoum/kibana/blob/be0a37522619c82800dc84cb1847aa88b658bb7b/x-pack/platform/packages/shared/onechat/onechat-genai-utils/tools/utils/resources/resolve_resource.ts#L38). Read more about it in this [slack thread](https://elastic.slack.com/archives/C08LX7YSHU2/p1761737404689789).


## How to test it?
Option 1 
* This PR includes automated agent tests that run on Phoenix. Please follow the instructions to run them:
[x-pack/solutions/security/packages/kbn-evals-suite-siem-entity-analytics/README.md](https://github.com/machadoum/kibana/blob/97c368e0ee41bd8ddf1e7d19f4802e708ea4a8b2/x-pack/solutions/security/packages/kbn-evals-suite-siem-entity-analytics/README.md)
* Add `gpt41Azure` connector to your `kibana.dev.yml`
* run `node scripts/playwright test --config x-pack/solutions/security/test/security_solution_evals/playwright.config.ts  --project="gpt41Azure"`

Option 2
* Enable the feature flag `securitySolution.naturalLanguageThreatHuntingEnabled`  and  `aiAssistant.aiAgents.enabled`
```yml
feature_flags.overrides:
  aiAssistant.aiAgents.enabled: true
  securitySolution.naturalLanguageThreatHuntingEnabled: true
```

* Add `gpt41Azure` connector to your `kibana.dev.yml`
* Create an agent using the UI exactly like the one in the automated test with the new entity analytics tool and prompt.
* Ask questions using the Agent Builder UI and selecting the created agent
* Make sure you have the necessary dependencies installed and data to answer your questions.

machadoum committed:
Create a siem agent and entity analytics tool

machadoum committed:
Improve prompts

machadoum committed:
Merge remote-tracking branch 'kibana/main' into siem-ea-14327-natural-language-entity-analytics

machadoum committed:
Small cleanup

machadoum committed:
Add space to tool

machadoum committed:
Add a security solution tool for getting the dataview indicies

machadoum committed:
Hacky implementation to make generate_query tool search multiple indices

machadoum committed:
Update prompt

machadoum committed:
Add ML jobs information to the security solution tool

machadoum committed:
Refactor and improve

kibanamachine commented:
<!-- kbn-message-context:cloud-deploy-job -->
Cloud deployment initiated, see credentials at: https://buildkite.com/elastic/kibana-deploy-cloud-from-pr/builds/541

machadoum committed:
Merge remote-tracking branch 'kibana/main' into siem-ea-14327-natural-language-entity-analytics

kibanamachine commented:
<!-- kbn-message-context:cloud-deploy-job -->
Cloud deployment initiated, see credentials at: https://buildkite.com/elastic/kibana-deploy-cloud-from-pr/builds/542

machadoum committed:
Increase ml node memory for testing

kibanamachine commented:
<!-- kbn-message-context:cloud-deploy-job -->
Cloud deployment initiated, see credentials at: https://buildkite.com/elastic/kibana-deploy-cloud-from-pr/builds/548

machadoum committed:
Improve response when feature is disabled

machadoum committed:
Increase ml node memory

machadoum committed:
Make several improvements to the threat hunting agent and tools

machadoum committed:
Small improvements

machadoum committed:
Fix ml node memory

machadoum committed:
feat: add evaluation suite for siem-entity-analytics agent

Add a new evaluation test suite package for the siem-entity-analytics agent builder.

Features:
- Chat client for Agent Builder API integration
- Criteria-based evaluation framework
- Sample tests for basic agent functionality and off-topic handling
- Simplified approach without HuggingFace dependencies

The suite includes:
- Basic security questions to verify agent role and capabilities
- Off-topic question handling to ensure agent stays focused on security

This follows the pattern from observability evaluations and integrates
with the kbn-evals framework for consistent LLM evaluation.

machadoum committed:
Wip: Add risk score tests

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/361242)
* Commit: 1d2c4eb5d093506a7ac619d28c26d8bad7f02818

### Failed CI Steps
* [Quick Checks](https://buildkite.com/elastic/kibana-pull-request/builds/361242#019a815b-b943-41f0-b46f-ed070afb4f99)
* [Checks](https://buildkite.com/elastic/kibana-pull-request/builds/361242#019a815b-b944-486f-85c9-911dc3f0b284)

### History
* :broken_heart: [Build #360154](https://buildkite.com/elastic/kibana-pull-request/builds/360154) failed 382bf3e341b7ce7be4e11cee2fb8426fee467c5d

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/361242","number":361242,"commit":"1d2c4eb5d093506a7ac619d28c26d8bad7f02818"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/360154","number":360154,"commit":"382bf3e341b7ce7be4e11cee2fb8426fee467c5d"}],"number":361242}
buildkite-pr-comment-->

machadoum committed:
Merge remote-tracking branch 'kibana/main' into siem-ea-14327-natural-language-entity-analytics

machadoum committed:
Add anomalies tests

machadoum committed:
Add capability to test tool calls and assert results

machadoum committed:
Update readme

machadoum committed:
Merge remote-tracking branch 'kibana/main' into siem-ea-14327-natural-language-entity-analytics

machadoum committed:
Fix CI issues

machadoum committed:
Code review improvements

machadoum committed:
Fix linter error

kibanamachine committed:
Changes from node scripts/lint_ts_projects --fix

kibanamachine committed:
Changes from node scripts/eslint_all_files --no-cache --fix

kibanamachine committed:
Changes from node scripts/generate codeowners

machadoum committed:
Fix some more coede review issues

kibanamachine committed:
Changes from node scripts/lint_ts_projects --fix

machadoum committed:
Improve dataview logic

machadoum committed:
Move security solution eval tests to security/test folder

machadoum committed:
Small prompt improvements

machadoum committed:
Fix duplicated security solution tsconfig

machadoum committed:
Fix import from restricted module

machadoum committed:
Fix code owners

machadoum committed:
Fix codeowner file

machadoum committed:
Create an evaluation test suit with several questions
and improve the agent and tools to answer them

machadoum committed:
Undo temporary changes

kibanamachine committed:
TO FIX: Run node 'scripts/yarn_deduplicate && yarn kbn bootstrap' locally, or add an exception to src/dev/yarn_deduplicate/index.ts and then commit the changes and push to your branch

kibanamachine committed:
Changes from security: 3rd-party dependencies

elasticmachine commented:
## :broken_heart: Build Failed
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/364735)
* Commit: f96050823103b4f489e001f29cd3ef70046e79ca

### Failed CI Steps
* [Quick Checks](https://buildkite.com/elastic/kibana-pull-request/builds/364735#019aba2c-d43d-42b0-ae51-7388635961c0)

### History
* :broken_heart: [Build #362675](https://buildkite.com/elastic/kibana-pull-request/builds/362675) failed 8c89a76f20380000e47e4fa8245bcb090cf64246
* :broken_heart: [Build #362665](https://buildkite.com/elastic/kibana-pull-request/builds/362665) failed 503645b11ed739a2be408b0e15ebac72166ecc3e
* :broken_heart: [Build #362642](https://buildkite.com/elastic/kibana-pull-request/builds/362642) failed c1c8046d85e46921f87ce76b60728f41ff67517a
* :broken_heart: [Build #362602](https://buildkite.com/elastic/kibana-pull-request/builds/362602) failed 6a21f2a731c03f07c774b2204494688061db1d23
* :broken_heart: [Build #362585](https://buildkite.com/elastic/kibana-pull-request/builds/362585) failed 8bb5e87097b4b27f436b55afefdd1221ee2a93b1

cc @machadoum

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/364735","number":364735,"commit":"f96050823103b4f489e001f29cd3ef70046e79ca"},{"buildStatus":{"state":"passed","success":false,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/362675","number":362675,"commit":"8c89a76f20380000e47e4fa8245bcb090cf64246"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/362665","number":362665,"commit":"503645b11ed739a2be408b0e15ebac72166ecc3e"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/362642","number":362642,"commit":"c1c8046d85e46921f87ce76b60728f41ff67517a"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/362602","number":362602,"commit":"6a21f2a731c03f07c774b2204494688061db1d23"}],"number":364735}
buildkite-pr-comment-->

machadoum committed:
Add experimental flag and rename tool and agent

machadoum committed:
Undo getPrivilegedMonitorUsersJoin moval

kibanamachine committed:
Changes from node scripts/regenerate_moon_projects.js --update

machadoum committed:
Update some imports

machadoum committed:
Remove .only from test

machadoum committed:
Merge remote-tracking branch 'kibana/main' into siem-ea-14327-natural-language-entity-analytics

machadoum committed:
Fix tool id

machadoum committed:
Improve tests

kibanamachine committed:
Changes from node scripts/lint_ts_projects --fix

kibanamachine committed:
Changes from node scripts/regenerate_moon_projects.js --update

machadoum committed:
Fix oops

machadoum committed:
Add security solution to internal namespace

jbudz reviewed (approved):
`x-pack/solutions/security/test/moon.yml` lgtm

machadoum committed:
Add 3 more test cases

machadoum committed:
Merge remote-tracking branch 'kibana/main' into siem-ea-14327-natural-language-entity-analytics

machadoum committed:
Delete entity analytics agent and update scout to use FF

machadoum committed:
Move entity analytics tool to another folder

machadoum committed:
Merge remote-tracking branch 'kibana/main' into siem-ea-14327-natural-language-entity-analytics

kibanamachine committed:
Changes from node scripts/lint_ts_projects --fix

tomsonpl reviewed (approved):
Defend workflows LGTM 👍

kibanamachine committed:
Changes from security: 3rd-party dependencies

kibanamachine committed:
Changes from node scripts/regenerate_moon_projects.js --update

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/374347)
* Commit: ba3c0cf1cee4e835ae023e07fa065c7a65b3c504

### Failed CI Steps
* [Quick Checks](https://buildkite.com/elastic/kibana-pull-request/builds/374347#019b2b9c-a37a-48f3-a75e-d99b5a78b781)

### History
* :broken_heart: [Build #374054](https://buildkite.com/elastic/kibana-pull-request/builds/374054) failed 65a509aa4eb3dfa12ed49247458db4450384a026
* :broken_heart: [Build #366210](https://buildkite.com/elastic/kibana-pull-request/builds/366210) failed 296a3f9ed983c5e8b30b5449fffe0efd57744dfc
* :broken_heart: [Build #365806](https://buildkite.com/elastic/kibana-pull-request/builds/365806) failed 7445e00243ab5fcf2661123c946ec340bf1f6cf1
* :broken_heart: [Build #365739](https://buildkite.com/elastic/kibana-pull-request/builds/365739) failed 82c572bf6760ab6643e774f4a4d5e0ed0cbb42fe
* :broken_heart: [Build #365011](https://buildkite.com/elastic/kibana-pull-request/builds/365011) failed d39a9bd6abb48b0c19d95ff723333f3ba5d75442
* :broken_heart: [Build #364735](https://buildkite.com/elastic/kibana-pull-request/builds/364735) failed f96050823103b4f489e001f29cd3ef70046e79ca

cc @machadoum

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/374347","number":374347,"commit":"ba3c0cf1cee4e835ae023e07fa065c7a65b3c504"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/374054","number":374054,"commit":"65a509aa4eb3dfa12ed49247458db4450384a026"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/366210","number":366210,"commit":"296a3f9ed983c5e8b30b5449fffe0efd57744dfc"},{"buildStatus":{"state":"running","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/365806","number":365806,"commit":"7445e00243ab5fcf2661123c946ec340bf1f6cf1"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/365739","number":365739,"commit":"82c572bf6760ab6643e774f4a4d5e0ed0cbb42fe"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/365011","number":365011,"commit":"d39a9bd6abb48b0c19d95ff723333f3ba5d75442"}],"number":374347}
buildkite-pr-comment-->

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/374362)
* Commit: e6250d27931eede5d3f63358b987836f6a8d7d07

### Failed CI Steps
* [Linting](https://buildkite.com/elastic/kibana-pull-request/builds/374362#019b2bb4-7ca7-4b8c-b7a2-d893b870ee12)

### History
* :broken_heart: [Build #374347](https://buildkite.com/elastic/kibana-pull-request/builds/374347) failed ba3c0cf1cee4e835ae023e07fa065c7a65b3c504
* :broken_heart: [Build #374054](https://buildkite.com/elastic/kibana-pull-request/builds/374054) failed 65a509aa4eb3dfa12ed49247458db4450384a026
* :broken_heart: [Build #366210](https://buildkite.com/elastic/kibana-pull-request/builds/366210) failed 296a3f9ed983c5e8b30b5449fffe0efd57744dfc
* :broken_heart: [Build #365806](https://buildkite.com/elastic/kibana-pull-request/builds/365806) failed 7445e00243ab5fcf2661123c946ec340bf1f6cf1
* :broken_heart: [Build #365739](https://buildkite.com/elastic/kibana-pull-request/builds/365739) failed 82c572bf6760ab6643e774f4a4d5e0ed0cbb42fe
* :broken_heart: [Build #365011](https://buildkite.com/elastic/kibana-pull-request/builds/365011) failed d39a9bd6abb48b0c19d95ff723333f3ba5d75442

cc @machadoum

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/374362","number":374362,"commit":"e6250d27931eede5d3f63358b987836f6a8d7d07"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/374347","number":374347,"commit":"ba3c0cf1cee4e835ae023e07fa065c7a65b3c504"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/374054","number":374054,"commit":"65a509aa4eb3dfa12ed49247458db4450384a026"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/366210","number":366210,"commit":"296a3f9ed983c5e8b30b5449fffe0efd57744dfc"},{"buildStatus":{"state":"running","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/365806","number":365806,"commit":"7445e00243ab5fcf2661123c946ec340bf1f6cf1"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/365739","number":365739,"commit":"82c572bf6760ab6643e774f4a4d5e0ed0cbb42fe"}],"number":374362}
buildkite-pr-comment-->

machadoum committed:
Fix CI

machadoum committed:
Rename folder back to the previous name

dmlemeshko reviewed (approved):
`src/platform/packages/shared/kbn-scout/src/servers/configs/custom/security_entity_analytics/stateful/stateful.config.ts` changes LGTM

machadoum committed:
Fix evaluation tests failing when running multiple at the same time

The converse API needs saved objects to work and the tests were deleting it.
If we need to delete saved objects, we must delete only the ones we need.
